### ISSUE TITLE
[BatchBundle] CountQueryBuilderOptimizer generate a wrong query

### ISSUE THREAD

**Issue by vtsykun** [oroinc/platform#778]
Step to reproduce
1) Branch 2.3 oro/crm oro/platform
2) Create datagrid with configuration that contains filter for relations `ManyToOne -> OneToMany -> ManyToOne`
Example of datagrid  - update cases-grid `Case->Contact->primaryAddress->region->name`. using the [patch](https://patch-diff.githubusercontent.com/raw/vtsykun/crm/pull/1.diff) 
```patch
diff --git a/src/Oro/Bundle/CaseBundle/Resources/config/oro/datagrids.yml b/src/Oro/Bundle/CaseBundle/Resources/config/oro/datagrids.yml
index 4b1c6c812f..2ee60f2ce3 100644
--- a/src/Oro/Bundle/CaseBundle/Resources/config/oro/datagrids.yml
+++ b/src/Oro/Bundle/CaseBundle/Resources/config/oro/datagrids.yml
@@ -170,6 +170,23 @@ datagrids:
                 link:         delete_link
     cases-grid:
         extends: base-case-grid
+        source:
+            query:
+                select:
+                    - contactRegion.name as contactRegionName
+                join:
+                    left:
+                        -
+                            join: relatedContact.addresses
+                            alias: contactAddress
+                            conditionType: WITH
+                            condition: "contactAddress.primary = true"
+                        -
+                            join: contactAddress.region
+                            alias: contactRegion
+        columns:
+            contactRegionName:
+                label: contactRegionName
         filters:
             columns:
                 relatedContact:
@@ -182,6 +199,10 @@ datagrids:
                     type: string
                     data_name: relatedAccountName
                     enabled: false
+                contactRegion:
+                    label: contactRegionName
+                    type: string
+                    data_name: contactRegionName
         options:
             entity_pagination: true
     account-cases-grid:
@@ -321,3 +342,4 @@ datagrids:
                         field_options:
                             class: OroCaseBundle:CasePriority
                             property: label
+
```
3) Go to cases index page and apply new filter
![selection_149](https://user-images.githubusercontent.com/21358010/33880353-6d27499e-df42-11e7-8500-6fb714f57639.png)
4)An error occurred during apply filter
![selection_148](https://user-images.githubusercontent.com/21358010/33880424-9db705cc-df42-11e7-964e-d4ae06cc0f00.png)

**Comment by NiMias** [comment]
Hello, @vtsykun.

Could you, please, also provide the minor version of your application (is it 2.3.9?) and the type of your DBMS (postgres, mysql?).

Because at the moment I can not reproduce this bug either in current 2.5 version nor in 2.3.9 version of OroCRM.

**Comment by vtsykun** [comment]
Hi @NiMias 
I can reproduce it on 2.5.0 and 2.3.10 on Postgres 9.5. using grid configuration https://github.com/vtsykun/crm/blob/45b15fbca4ecf31ccd0ed6624e9cd1a452959851/src/Oro/Bundle/CaseBundle/Resources/config/oro/datagrids.yml
Also I can provide [demo instance](http://dev.de1.yandex.ovh/datagrid/cases-grid?appearanceType=grid&cases-grid%5B_pager%5D%5B_page%5D=1&cases-grid%5B_pager%5D%5B_per_page%5D=25&cases-grid%5B_parameters%5D%5Brefresh%5D=true&cases-grid%5B_parameters%5D%5Bview%5D=__all__&cases-grid%5B_appearance%5D%5B_type%5D=grid&cases-grid%5B_sort_by%5D%5BreportedAt%5D=DESC&cases-grid%5B_filter%5D%5BcontactRegion%5D%5Bvalue%5D=test&cases-grid%5B_filter%5D%5BcontactRegion%5D%5Btype%5D=1)

**Comment by vtsykun** [comment]
The problem is related to [listener](https://github.com/oroinc/platform/blob/master/src/Oro/Bundle/FormBundle/EventListener/CountQueryOptimizationListener.php#L13) that remove all "primary" joins (joins that match to regexp `/%s.primary\s*=\s*(?P<value>true|1|:\w+|\?\d+)/`) 
In some cases, as in my [example](https://github.com/vtsykun/crm/blob/45b15fbca4ecf31ccd0ed6624e9cd1a452959851/src/Oro/Bundle/CaseBundle/Resources/config/oro/datagrids.yml#L171-L207), it leads to errors
possible fix: #789

**Comment by manowark** [comment]
Hello, @vtsykun.
Please, sing our Contributor License Agreement. See https://github.com/oroinc/platform/pull/789

**Comment by aivus** [comment]
Fixed in #789 and will appear in master after sync
